src/theory/sets/cardinality_extension.cpp

```diff
@@ -0,0 +1,911 @@
+/*********************                                                        */
+/*! \file cardinality_extension.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of an extension of the theory sets for handling
+ ** cardinality constraints.
+ **/
+
+#include "theory/sets/cardinality_extension.h"
+
+#include "expr/emptyset.h"
+#include "expr/node_algorithm.h"
+#include "options/sets_options.h"
+#include "theory/sets/normal_form.h"
+#include "theory/valuation.h"
+
+using namespace std;
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace sets {
+
+CardinalityExtension::CardinalityExtension(SolverState& s,
+                                           InferenceManager& im,
+                                           eq::EqualityEngine& e,
+                                           context::Context* c,
+                                           context::UserContext* u)
+    : d_state(s), d_im(im), d_ee(e), d_card_processed(u)
+{
+  d_zero = NodeManager::currentNM()->mkConst(Rational(0));
+  // we do congruence over cardinality
+  d_ee.addFunctionKind(CARD);
+}
+
+void CardinalityExtension::reset()
+{
+  d_eqc_to_card_term.clear();
+  d_t_card_enabled.clear();
+}
+void CardinalityExtension::registerTerm(Node n)
+{
+  Trace("sets-card-debug") << "Register term : " << n << std::endl;
+  Assert(n.getKind() == CARD);
+  TypeNode tnc = n[0].getType().getSetElementType();
+  d_t_card_enabled[tnc] = true;
+  Node r = d_ee.getRepresentative(n[0]);
+  if (d_eqc_to_card_term.find(r) == d_eqc_to_card_term.end())
+  {
+    d_eqc_to_card_term[r] = n;
+    registerCardinalityTerm(n[0]);
+  }
+  if (tnc.isInterpretedFinite())
+  {
+    std::stringstream ss;
+    ss << "ERROR: cannot use cardinality on sets with finite element "
+          "type (term is "
+       << n << ")." << std::endl;
+    throw LogicException(ss.str());
+    // TODO (#1123): extend approach for this case
```

Comment:
Interesting. I didn't know the cardinality in the example below is not supported. 
```
(set-logic ALL_SUPPORTED)
(set-info :status unsat)
(declare-fun S () (Set Bool))
(declare-fun x () Int)
(assert (= (card S) 3))
(check-sat)
```
```
(error "ERROR: cannot use cardinality on sets with finite element type (term is (card S)).")
```

Reply:
Yep, see https://github.com/CVC4/CVC4/issues/1123.  We should implement support for this.